hist/hist/src/THnChain.cxx

```diff
@@ -0,0 +1,216 @@
+ * Copyright (C) 1995-2016, Rene Brun and Fons Rademakers.               *
+ * All rights reserved.                                                  *
+ *                                                                       *
+ * For the licensing terms see $ROOTSYS/LICENSE.                         *
+ * For the list of contributors see $ROOTSYS/README/CREDITS.             *
+ *************************************************************************/
+
+#include "THnChain.h"
+
+#include "TArray.h"
+#include "TAxis.h"
+#include "TDirectory.h"
+#include "TFile.h"
+#include "TH1.h"
+#include "TH2.h"
+#include "TH3.h"
+#include "THnBase.h"
+#include "TMath.h"
+#include "TMetaUtils.h"
+
+void THnChain::AddFile(const char* fileName)
+{
+   fFiles.emplace_back(fileName);
+
+   // Initialize axes from first seen instance.
+   if (fAxes.empty()) {
+      THnBase* hs = ReadHistogram(fileName);
+
+      if (hs) {
+         const Int_t naxes = hs->GetNdimensions();
+         for (Int_t i = 0; i < naxes; ++i) {
+            fAxes.push_back(hs->GetAxis(i));
+         }
+      }
+   }
+}
+
+TAxis* THnChain::GetAxis(Int_t i)
+{
+   if (i < 0 || i >= static_cast<Int_t>(fAxes.size())) {
+      return nullptr;
+   }
+
+   return fAxes[i];
+}
+
+TObject* THnChain::ProjectionAny(Int_t ndim, const Int_t* dim, Option_t* option)
+{
+   if (ndim == 0) {
+      return nullptr;
+   }
+
+   TObject* h_merged = nullptr;
+   for (const auto& file : fFiles) {
+      THnBase* hs = ReadHistogram(file.c_str());
+
+      if (!hs) {
+         Warning("ProjectionAny",
+                 "Could not find histogram %s in file %s",
+                 fName.c_str(),
+                 file.c_str());
+
+         continue;
+      }
+
+      if (!CheckConsistency(*hs, fAxes)) {
+         Warning("ProjectionAny",
+                 "Histogram %s from file %s is inconsistent with the histogram from file %s",
+                 fName.c_str(),
+                 file.c_str(),
+                 fFiles[0].c_str());
+
+         continue;
+      }
+
+      SetupAxes(*hs);
+
+      // Perform projection.
+      TObject* h = nullptr;
+
+      if (ndim == 1) {
+         h = hs->Projection(dim[0], option);
+      } else if (ndim == 2) {
+         h = hs->Projection(dim[0], dim[1], option);
+      } else if (ndim == 3) {
+         h = hs->Projection(dim[0], dim[1], dim[2], option);
+      } else {
+         h = hs->ProjectionND(ndim, dim, option);
+      }
+
+      delete hs;
+
+      // Add this histogram.
+      if (h_merged) {
+         if (ndim < 3) {
+            static_cast<TH1*>(h_merged)->Add(static_cast<TH1*>(h));
+         } else {
+            static_cast<THnBase*>(h_merged)->Add(static_cast<THnBase*>(h));
+         }
+
+         delete h;
+      } else {
+         h_merged = h;
+      }
+   }
+
+   return h_merged;
+}
+
+THnBase* THnChain::ReadHistogram(const char* fileName)
+{
+   TFile* f = TFile::Open(fileName);
+
+   if (!f) {
+      return nullptr;
+   }
+
+   TDirectory::TContext ctxt(f);
```

Comment:
This should be constructed before the call to `TFile::Open`, to capture the previous value of `gDirectory`.